### ISSUE TITLE
ohai reload resets Chef::Log.level

### ISSUE THREAD

**Issue by jaymzh** [chef/chef#6734]
## Description

If you dynamically set the log level in your run using `Chef::Log.level = :debug`, and then later notify the `ohai` resource, the loglevel will be reset to `info`.

## Chef Version
12.21.26

## Platform Version
CentOS 7

## Replication Case
Using this recipe:

```
ohai 'etc' do
    plugin 'etc'
    action :nothing
end
Chef::Log.level = :debug
log 'test 1' do
    level :debug
    notifies :reload, 'ohai[etc]', :immediately
end
log 'test 2' do
    level :debug
end 
```

You can see the log level resets after the ohai reload:

```
[2018-01-04T17:03:46-08:00] INFO: Processing ohai[etc] action nothing (fb_util::test line 1)
[2018-01-04T17:03:46-08:00] DEBUG: Skipping ohai[etc] due to action :nothing
[2018-01-04T17:03:46-08:00] INFO: Processing log[test 1] action write (fb_util::test line 6)
[2018-01-04T17:03:46-08:00] DEBUG: Providers for generic log resource enabled on node include: [Chef::Provider::Log::ChefLog]
[2018-01-04T17:03:46-08:00] DEBUG: Provider for action write on resource log[test 1] is Chef::Provider::Log::ChefLog
[2018-01-04T17:03:46-08:00] DEBUG: test 1
[2018-01-04T17:03:46-08:00] INFO: log[test 1] sending reload action to ohai[etc] (immediate)
[2018-01-04T17:03:46-08:00] INFO: Processing ohai[etc] action reload (fb_util::test line 1)
[2018-01-04T17:03:46-08:00] DEBUG: Providers for generic ohai resource enabled on node include: [Chef::Provider::Ohai]
[2018-01-04T17:03:46-08:00] DEBUG: Provider for action reload on resource ohai[etc] is Chef::Provider::Ohai
[2018-01-04T17:03:46-08:00] DEBUG: Running Ohai with the following configuration: ***REDACTED***
[2018-01-04T17:03:46-08:00] INFO: The plugin path /etc/chef/ohai/plugins does not exist. Skipping...
[2018-01-04T17:03:46-08:00] WARN: Plugin Definition Error: </etc/chef/ohai_plugins/shard.rb>: collect_data already defined on platform darwin
[2018-01-04T17:03:46-08:00] WARN: Plugin Definition Error: </etc/chef/ohai_plugins/linux/hostnamectl.rb>: collect_data already defined on platform linux
[2018-01-04T17:03:46-08:00] WARN: Plugin Definition Error: </etc/chef/ohai_plugins/linux/machineid.rb>: collect_data already defined on platform linux
[2018-01-04T17:03:46-08:00] INFO: [ohai] Loading HP BIOS plugin
[2018-01-04T17:03:48-08:00] INFO: ohai[etc] reloaded
[2018-01-04T17:03:48-08:00] INFO: Processing log[test 2] action write (fb_util::test line 10)
[2018-01-04T17:03:48-08:00] WARN: Skipping final node save because override_runlist was given
[2018-01-04T17:03:48-08:00] INFO: Chef Run complete in 3.176964345 seconds
```

## Client Output
See above.

## Stacktrace
N/A

**Comment by jaymzh** [comment]
It's worth nothing this wasn't the case in 12.13.37

**Comment by btm** [comment]
Bringing this line back fixes this: https://github.com/chef/ohai/pull/942/files#diff-dee5766eb9ab7555535dae03a4674f28L225

**Comment by btm** [comment]
@thommay due to https://github.com/chef/ohai/pull/942 we now inherit Chef's logger, and then we change the log level in `Ohai::System.configure_logging`. We previously created a new logger so that change didn't matter. We don't want to be changing the log level of chef's logger with ohai's log level [and not changing it back].

The only way I can see to know in `configure_logging` that we have Chef's logger is that the class is MonoLogger. That class is unique to Chef, but that's a whole pile of crazy.

Could we bring back the `Ohai::Log.init` as a duplicate (keep it in application.rb too)? That'd be simple in a world full of edge cases, but would that create any?

**Comment by jaymzh** [comment]
One of the things I *think* that fixed was the fact that in plugins, Ohai::Log.info() messages would only show up in ohai CLI runs, but if you wanted it to show up in Chef runs you'd have to do Chef::Log.info(), and it seemed silly to have to do both... but I could be wrong.

**Comment by jaymzh** [comment]
So, @btm and @thommay, if I throw this into a library file, it "fixes" it:

```
module Ohai
  class System
    def configure_logging
      return if Ohai::Log.configured?
      Ohai::Log.init(Ohai.config[:log_location])
      if Ohai.config[:log_level] == :auto
        Ohai::Log.level = :info
      else
        Ohai::Log.level = Ohai.config[:log_level]
      end
    end
  end
end
```

But I have to admit the way in which the loggers play together isn't something I fully understand... is this going to cause any other problems?

**Comment by tas50** [comment]
@jaymzh Does that still work correctly when you run ohai from the command line and specify logging there. That was the previous issue we fixed which introduce the current regression.

**Comment by jaymzh** [comment]
@tas50 - that code can't possibly effect Ohai from the command line since it's in a cookbook library file. :)

However what I imagine it *might* effect is the level of plugin logging you get inside of the Chef run...

**Comment by tas50** [comment]
@jaymzh I put that same change into a Ohai and it does break the debug logging from the CLI again so we'll have to come up with another fix there.

**Comment by btm** [comment]
`return if Ohai::Log.configured?` should always return there in this case because we've run `Ohai::Log.init` already in application.rb if we're running from the command line, and we've already initialized a logger in Chef if we're coming through `Ohai::System.new`.

I would guess adding only `Ohai::Log.init(Ohai.config[:log_location])` back to configure_logging would "work" in both cases, but it wouldn't be Right.

**Comment by jaymzh** [comment]
@tas50 @btm  - yeah I dont' think it's a change that's right for Ohai... but I *think* it'll work if I'm only monkeypatching Ohai from within a chef run...  I think...

**Comment by jaymzh** [comment]
i.e. I'm looking for a short-term work around :)

**Comment by thommay** [comment]
@jaymzh yeah just drop that monkey patch into a cookbook. We'll fix it properly in Chef 14 but for right now that's the correct thing to do.

**Comment by jaymzh** [comment]
Did it yesterday. :)

We'll need something less hacky for 13 though... not 14. No need to backport to 12, we'll be off of it soon.

**Comment by thommay** [comment]
since a work around exists, and this doesn't affect the correctness of Chef, 
we're not going to work on this in 13, but will ship a correct fix in 14. Fixing this in 13 will require deep surgery that simply isn't appropriate to where in it's lifecycle 13 is.

**Comment by jaymzh** [comment]
Chef 14 isn't out until April. So you're not shipping a fix for a pretty terrible bug for 3 more months?

That "work around" is monkeypatching functions that may change, and only works because it *happens* to get sucked in to late to do damage. Meaning that such a "work around" is dangerous over even the mid-term much less the long-term.

I understand not wanting to do surgery to 13... but we have backported entire rewrites of attributes to 12, so lets not rule it out just yet. :) Once we understand the final solution (do we even have a draft of a PR yet? I feel like we're talking hypotheticals) we can figure out what that looks like... perhaps there's a less-invasive but less-hacky work around we can bring to 13.

To be clear. This was broken somewhere between 12.13 and 12.21. Chef 14 isn't out until April and Chef 13 is supported until 2019. It doesn't seem reasonable to leave basically *all* users broken for the next 3 months and then the vast majority of users for the next year until they get past other non-back-compat changes to move to 14.

**Comment by thommay** [comment]
Lets be clear: this is a bug that has existed for well over a year and has been reported by a single customer. Almost no-one in the wild does this, and that's why I'm comfortable saying it'll be fixed in 14. (There are also very few incompatible changes from 13 to 14).

**Comment by lamont-granquist** [comment]
didn't i fix this in https://github.com/chef/ohai/pull/955 ?

running on chef-13.7.13 / ohai-13.7.1 i get:

```
[2018-01-18T11:59:08-08:00] DEBUG: test 1
```

the key line from my patch is that configure_logging is not called if we're coming from chef-client and not the command line:

```
configure_logging if @cli
```

**Comment by jaymzh** [comment]
Thanks @lamont-granquist . I just setup a box with 13.6.4 for Debian, downloaded from chef.io...

I can confirm that this is fixed. However, someone may care that it double-prints every single line, even:
```
[2018-01-18T22:04:20-08:00] INFO: *** Chef 13.6.4 ***
[2018-01-18T22:04:20-08:00] INFO: *** Chef 13.6.4 ***
[2018-01-18T22:04:20-08:00] INFO: Platform: x86_64-linux
[2018-01-18T22:04:20-08:00] INFO: Platform: x86_64-linux
[2018-01-18T22:04:20-08:00] INFO: Chef-client pid: 5528
[2018-01-18T22:04:20-08:00] INFO: Chef-client pid: 5528
```

Probably something wonky with that build. But the loglevel is definitely not reset. I don't have a normal chef 13 env to compare to, this was just a simple chef-zero/chef-client setup I made at home to test this. Probably not related to this bug, but ... figured I'd mention it.

As always, @lamont-granquist is my hero. Fixed in 13, sans surgery. :)
